diff --git a/third_party/systemlibs/grpc.bazel.grpc_deps.bzl b/third_party/systemlibs/grpc.bazel.grpc_deps.bzl
index 97f7db0..21af1a4 100644
--- a/third_party/systemlibs/grpc.bazel.grpc_deps.bzl
+++ b/third_party/systemlibs/grpc.bazel.grpc_deps.bzl
@@ -1,7 +1,19 @@
 """Load dependencies needed to compile and test the grpc library as a 3rd-party consumer."""
 
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
 def grpc_deps():
     """Loads dependencies need to compile and test the grpc library."""
 
-    pass
+    native.bind(
+        name = "upb_lib",
+        actual = "@upb//:upb",
+    )
 
+    if "upb" not in native.existing_rules():
+        http_archive(
+            name = "upb",
+            sha256 = "61d0417abd60e65ed589c9deee7c124fe76a4106831f6ad39464e1525cef1454",
+            strip_prefix = "upb-9effcbcb27f0a665f9f345030188c0b291e32482",
+            url = "https://github.com/protocolbuffers/upb/archive/9effcbcb27f0a665f9f345030188c0b291e32482.tar.gz",
+        )
